File: wp-plugins/riseup-asia-uploader/includes/Enums/ActionType.php
<?php
/**
 * ActionType — Transaction logging action identifiers.
 *
 * @package RiseupAsia\Enums
 * @since   2.1.0
 */

namespace RiseupAsia\Enums;

if (!defined('ABSPATH')) {
    exit;
}

enum ActionType: string
{
    // Core plugin actions
    case Upload           = 'Upload';
    case UploadActive     = 'UploadActive';
    case UploadInitiated  = 'UploadInitiated';
    case Enable           = 'Enable';
    case Disable          = 'Disable';
    case Delete           = 'Delete';
    case FileReplace      = 'FileReplace';
    case FileDelete       = 'FileDelete';
    case Sync             = 'Sync';
    case SyncDelete       = 'SyncDelete';

    // Post/content actions
    case PostCreate       = 'PostCreate';
    case PostUpdate       = 'PostUpdate';
    case CategoryCreate   = 'CategoryCreate';
    case MediaUpload      = 'MediaUpload';

    // Auth
    case AuthFailed       = 'AuthFailed';

    // Export actions
    case ExportSelf       = 'ExportSelf';
    case ExportPlugin     = 'ExportPlugin';

    // Agent actions
    case AgentAdd           = 'AgentAdd';
    case AgentRemove        = 'AgentRemove';
    case AgentTest          = 'AgentTest';
    case AgentSync          = 'AgentSync';
    case AgentApiError      = 'AgentApiError';

    // Snapshot actions
    case SnapshotCreate          = 'SnapshotCreate';
    case SnapshotRestore         = 'SnapshotRestore';
    case SnapshotDelete          = 'SnapshotDelete';
    case SnapshotExport          = 'SnapshotExport';
    case SnapshotImport          = 'SnapshotImport';
    case SnapshotCleanup         = 'SnapshotCleanup';
    case SnapshotFullBackup      = 'SnapshotFullBackup';
    case SnapshotIncremental     = 'SnapshotIncremental';
    case SnapshotSettingsUpdate  = 'SnapshotSettingsUpdate';
    case SnapshotZipBuild        = 'SnapshotZipBuild';
    case SnapshotZipExpire       = 'SnapshotZipExpire';
    case SnapshotZipDownload     = 'SnapshotZipDownload';

    public function isEqual(self $other): bool { return $this === $other; }
    public function isOtherThan(self $other): bool { return $this !== $other; }
    public function isAnyOf(self ...$others): bool { return in_array($this, $others, true); }

    public function isSnapshot(): bool { return str_starts_with($this->value, 'Snapshot'); }
    public function isAgent(): bool    { return str_starts_with($this->value, 'Agent'); }
    

    public function isLifecycle(): bool
    {
        return $this->isAnyOf(self::Enable, self::Disable, self::Delete);
    }
}

File: backend/internal/wordpress/action_type.go
// Package wordpress provides typed action constants for transaction logging.
//
// ActionType replaces untyped string constants for action identifiers.
// Each constant matches the PHP ActionType enum values for cross-language consistency.
package wordpress

import "strings"

// ActionType is a typed string representing a transaction logging action.
type ActionType string

// =============================================================================
// CORE PLUGIN ACTIONS
// =============================================================================

const (
	// ActionUpload represents a plugin upload action.
	ActionUpload ActionType = "upload"

	// ActionUploadActive represents upload + activate action.
	ActionUploadActive ActionType = "upload_active"

	// ActionEnable represents a plugin activation action.
	ActionEnable ActionType = "enable"

	// ActionDisable represents a plugin deactivation action.
	ActionDisable ActionType = "disable"

	// ActionDelete represents a plugin deletion action.
	ActionDelete ActionType = "delete"

	// ActionFileReplace represents a single file replacement action.
	ActionFileReplace ActionType = "file_replace"

	// ActionFileDelete represents a single file deletion action.
	ActionFileDelete ActionType = "file_delete"

	// ActionSync represents a delta file sync action.
	ActionSync ActionType = "sync"
)

// =============================================================================
// POST/CONTENT ACTIONS
// =============================================================================

const (
	// ActionPostCreate represents a blog post creation action.
	ActionPostCreate ActionType = "post_create"

	// ActionPostUpdate represents a blog post update action.
	ActionPostUpdate ActionType = "post_update"

	// ActionCategoryCreate represents a category creation action.
	ActionCategoryCreate ActionType = "category_create"

	// ActionMediaUpload represents a media library upload action.
	ActionMediaUpload ActionType = "media_upload"
)

// =============================================================================
// AUTH ACTIONS
// =============================================================================

const (
	// ActionAuthFailed represents an authentication failure.
	ActionAuthFailed ActionType = "auth_failed"
)

// =============================================================================
// EXPORT ACTIONS
// =============================================================================

const (
	// ActionExportSelf represents exporting the plugin itself.
	ActionExportSelf ActionType = "export_self"

	// ActionExportPlugin represents exporting an arbitrary plugin as ZIP.
	ActionExportPlugin ActionType = "export_plugin"
)

// IsEqual checks if this action equals the given action.
func (a ActionType) IsEqual(other ActionType) bool {
	return a == other
}

// String returns the raw string value of the action.
func (a ActionType) String() string {
	return string(a)
}

// IsSnapshot checks if this is a snapshot-related action.
func (a ActionType) IsSnapshot() bool {
	return strings.HasPrefix(string(a), "snapshot_")
}

// IsAgent checks if this is an agent-related action.
func (a ActionType) IsAgent() bool {
	return strings.HasPrefix(string(a), "agent_")
}

// IsUpdate checks if this is an update-related action.
func (a ActionType) IsUpdate() bool {
	return strings.HasPrefix(string(a), "update_")
}

// IsLifecycle checks if this is a plugin lifecycle action (enable/disable/delete).
func (a ActionType) IsLifecycle() bool {
	return a.IsEqual(ActionEnable) || a.IsEqual(ActionDisable) || a.IsEqual(ActionDelete)
}
